pkg/cli/init_project.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"sigs.k8s.io/kubebuilder/pkg/cli/internal"
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+)
+
+func (c cli) newInitProjectCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "Initialize a new project",
+		Long: `Initialize a new project.
+
+For further help about a specific project version, set --project-version.
+`,
+		Run:     func(cmd *cobra.Command, args []string) {},
+		Example: c.getInitHelpExamples(),
+	}
+
+	// Register --project-version on the dynamically created command
+	// so that it shows up in help and does not cause a parse error.
+	cmd.Flags().String("project-version", c.defProjVersion,
+		fmt.Sprintf("project version, (%s)", strings.Join(c.getAvailableProjectVersions(), ", ")))
+
+	// Pre-parse the project version and help flags so that we can
+	// dynamically bind to a plugin's init implementation (or not).
+	projectVersion, isHelpOnly := c.getBaseFlags()
+
+	// If only the help flag was set, return the command as is.
+	if isHelpOnly {
+		return cmd
+	}
+
+	// Lookup the plugin for projectVersion and bind it to the command.
+	c.bindInit(cmd, projectVersion)
+	return cmd
+}
+
+func (c cli) getInitHelpExamples() string {
+	vs := c.getAvailableProjectVersions()
+	var sb strings.Builder
+
+	for _, v := range vs {
+		rendered := fmt.Sprintf(`  # Help for initializing a project with version %s
+  %s init --project-version=%s -h
+
+`, v, c.commandName, v)
+		sb.WriteString(rendered)
+	}
+	return strings.TrimSuffix(sb.String(), "\n\n")
+}
+
+func (c cli) getAvailableProjectVersions() (projectVersions []string) {
+	for _, ps := range c.plugins {
+		for _, p := range ps {
+			// Only return project versions from non-deprecated plugins
+			if _, ok := p.(plugin.Deprecated); ok {
+				continue
+			}
+			if _, ok := p.(plugin.Init); !ok {
+				continue
+			}
+			projectVersions = append(projectVersions, strconv.Quote(p.Version()))
+		}
+	}
+	return projectVersions
+}
```

Comment:
I think you should break from the internal loop once you add a version to the slice.
Why use `p.Version()` if the outter loop already has the version as key?
Additionally I suggest some new variable names to make them more descriptive:

```go
func (c cli) getAvailableProjectVersions() (versions []string) {
	for version, plugins := range c.plugins {
		for _, p := range plugins {
			// Only return project versions from non-deprecated plugins
			if _, deprecated := p.(plugin.Deprecated); deprecated {
				continue
			}
			if _, isInitPlugin := p.(plugin.Init); !isInitPlugin {
				continue
			}
			versions = append(versions, strconv.Quote(version))
			break
		}
	}
	return
}
```

Also, is `strconv.Quote` needed?

Reply:
I made an update to this method that got lost in the rebase. Will update.